### PR TITLE
[LiveRangeEdit] Remove any MemoryOperand on MI when converting it to KILL.

### PR DESCRIPTION
**Pull request by JonPsson1**
An optimization pass may change an instruction into a KILL it seems when there is no use for the produced value. In this case the register coalescer did so while leaving the MMO untouched. The MachineVerifier did not accept this, but it seems reasonable that it should.

This patch makes the machine verifier accept dangling MMOs on KILL instructions.

The alternative might be to have the actor always call something like MI->removeAsKill()  which also removed the MMO after setting the opcode to KILL.

@arsenm @ellishg @topperc @michaelmaitland

### CODE REVIEW

File: llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.ll [hunk 1]
```diff
@@ -0,0 +1,30 @@
+; RUN: llc < %s -mtriple=s390x-linux-gnu -mcpu=z16 -disable-machine-dce \
+; RUN:   -verify-machineinstrs -O3
+;
+; Test that this passes the verifier even though a KILL instruction with a
+; dangling memory operand emerges.
+
+define void @fun(ptr %Src, ptr %Dst) {
+  br label %2
+
+2:
```
Comment (arsenm) [inline llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.ll:10]: Use named values in tests

**Comment by arsenm** [comment]
> An optimization pass may change an instruction into a KILL it seems when there is no use for the produced value. In this case the register coalescer did so while leaving the MMO untouched. The MachineVerifier did not accept this, but it seems reasonable that it should.


It should just delete the memory operands. There's no reason to accept this

File: llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.ll [hunk 2]
```diff
@@ -0,0 +1,30 @@
+; RUN: llc < %s -mtriple=s390x-linux-gnu -mcpu=z16 -disable-machine-dce \
+; RUN:   -verify-machineinstrs -O3
```
Comment (arsenm) [inline llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.ll:2]: Probably don't need -O3

**Review by arsenm — COMMENTED** [review]
Probably need a more targeted mir test

**Comment by JonPsson1** [comment]
Updated per review.

> It should just delete the memory operands. There's no reason to accept this

ok - I thought it was ok to make things simple for the CodeGen passes and just let them switch to KILL opcode, but I see now that they actually take care to remove operands and MMOs.

File: llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.mir [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  - { id: 4, class: gr64bit }
+  - { id: 5, class: grx32bit }
+  - { id: 6, class: grx32bit }
+  - { id: 7, class: grx32bit }
+  - { id: 8, class: addr64bit }
+liveins:
+  - { reg: '$r2d', virtual-reg: '%3' }
+frameInfo:
+  maxAlignment:    1
+machineFunctionInfo: {}
```
Comment (arsenm) [inline llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.mir:34]:
```suggestion
tracksRegLiveness: true
```

File: llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.mir [hunk 4]
```diff
@@ -0,0 +1,52 @@
+# RUN: llc -o /dev/null %s -mtriple=s390x-linux-gnu -mcpu=z16 \
+# RUN:   -verify-machineinstrs -run-pass=register-coalescer
+
+# Test that the MemoryOperand of the produced KILL instruction is removed
+# and as a result the machine verifier succeeds.
+
+--- |
```
Comment (arsenm) [inline llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.mir:7]: Don't need IR section

File: llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.mir [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    liveins: $r2d
+  
+    %3:addr64bit = COPY killed $r2d
+  
+  bb.1:
+    %5:grx32bit = LMux killed %3, 0, $noreg :: (load (s32) from %ir.Src)
+    CHIMux killed %5, 0, implicit-def $cc
+    %7:grx32bit = LHIMux 0
+    %1:grx32bit = COPY killed %7
+    %1:grx32bit = LOCMux %1, undef %8:addr64bit, 0, 14, 6, implicit killed $cc :: (load (s32) from %ir.Src2)
```
Comment (arsenm) [inline llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.mir:46]:
```suggestion
    %1:grx32bit = LOCMux %1, undef %8:addr64bit, 0, 14, 6, implicit killed $cc :: (load (s32))
```

**Review by arsenm — COMMENTED** [review]
Description and test name need update

File: llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.mir (hunk 3 above)
Comment (JonPsson1) [reply llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.mir:34]: actually, that line is already present further up...

**Comment by JonPsson1** [comment]
test updated

File: llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.mir (hunk 3 above)
Comment (arsenm) [reply llvm/test/CodeGen/SystemZ/machine-ver-kill-memop.mir]: This is to delete everything else, not add the line

**Review by arsenm — COMMENTED** [review]
Test name and description still does not reflect current patch

**Comment by JonPsson1** [comment]
One more try with the test :)

**Review by arsenm — COMMENTED** [review]
Patch title is still wrong

**Comment by JonPsson1** [comment]
> Patch title is still wrong

I rewrote it here now, and I will write a better commit message as well when I merge it.

**Review by arsenm — APPROVED** [review]

**Comment by JonPsson1** [comment]
Committed - thanks for review!

**Pull request merged**